expose.go

```diff
@@ -0,0 +1,160 @@
+package chromedp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/chromedp/cdproto/page"
+	"github.com/chromedp/cdproto/runtime"
+)
+
+// BindingCalledPayload ...
+type BindingCalledPayload struct {
+	Type string `json:"type"`
+	Name string `json:"name"`
+	Seq  int64  `json:"seq"`
+	Args string `json:"args"`
+}
+
+// BindingFunc expose function type
+type BindingFunc func(args string) (string, error)
+
+// ExposeFunc The method adds a function called name on the browser page's window object.
+// When called, the function executes BindingFunc in go env
+// and returns a Promise which resolves to the return value of BindingFunc.
+// Note. compared with puppeteer's exposeFunction.
+// the BindingFunc takes exactly one argument, this argument should be string
+func ExposeFunc(ctx context.Context, fnName string, fn BindingFunc) error {
+	c := FromContext(ctx)
+	if c == nil {
+		return ErrInvalidContext
+	}
+	if c.Target == nil {
+		if err := c.newTarget(ctx); err != nil {
+			return err
+		}
+	}
+
+	c.Target.bindingFuncListenOnce.Do(func() {
+		c.Target.bindingFuncs = make(map[string]BindingFunc)
+
+		err := Run(ctx, ActionFunc(func(ctx context.Context) error {
+			_, err := page.AddScriptToEvaluateOnNewDocument(exposedFunJS).Do(ctx)
+			return err
+		}))
+		if err != nil {
+			return
+		}
+
+		ListenTarget(ctx, func(ev interface{}) {
+			switch ev := ev.(type) {
+			case *runtime.EventBindingCalled:
+				var payload BindingCalledPayload
+
+				err := json.Unmarshal([]byte(ev.Payload), &payload)
+				if err != nil {
+					return
+				}
+
+				var expression string
+
+				c.Target.bindingFuncMu.RLock()
+				defer c.Target.bindingFuncMu.RUnlock()
+				if fn, ok := c.Target.bindingFuncs[payload.Name]; ok {
+					res, err := fn(payload.Args)
+					if err != nil {
+						expression = deliverError(payload.Name, payload.Seq, err.Error(), err.Error())
+					} else {
+						expression = deliverResult(payload.Name, payload.Seq, res)
+					}
+				} else {
+					expression = deliverError(payload.Name, payload.Seq, "bindingCall name not exsit", "")
+				}
+
+				go func() {
+					Run(ctx, Evaluate(expression, nil, func(p *runtime.EvaluateParams) *runtime.EvaluateParams {
+						return p.WithContextID(ev.ExecutionContextID)
+					}))
+				}()
+
+			}
+		})
+	})
+
+	c.Target.bindingFuncMu.Lock()
+	if _, ok := c.Target.bindingFuncs[fnName]; ok {
+		c.Target.bindingFuncMu.Unlock()
+		return ErrExposeNameExist
+	}
+	c.Target.bindingFuncs[fnName] = fn
+	c.Target.bindingFuncMu.Unlock()
+
+	err := Run(ctx, runtime.AddBinding(fnName))
+	if err != nil {
+		return err
+	}
+
+	expression := addPageBinding("exposedFun", fnName)
+	err = Run(ctx, ActionFunc(func(ctx context.Context) error {
+		_, err := page.AddScriptToEvaluateOnNewDocument(expression).Do(ctx)
+		return err
+	}))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+const exposedFunJS = `
```

Comment:
Extrat the js code into a standalone file and save it to the `js` folder. Then load the js code with `//go:embed`.